File: snippets/cpp/geometry/polygon.cpp
ld polygon_area(vector<Pt> &p) {
    ld a = 0; int n = sza(p);
    for (int i = 0; i < n; i++)
        a += p[i].cross(p[(i + 1) % n]);
    return abs(a) / 2;
}
bool in_polygon(vector<Pt> &p, Pt q) {
    int n = sza(p), cnt = 0;
    for (int i = 0; i < n; i++) {
        Pt a = p[i], b = p[(i + 1) % n];
        if ((a.y <= q.y && q.y < b.y) || (b.y <= q.y && q.y < a.y))
            if (q.x < a.x + (b.x - a.x) * (q.y - a.y) / (b.y - a.y))
                cnt++;
    }
    return cnt & 1;
}

File: snippets/rust/geometry/polygon.rs
// Requires the `point` snippet.
fn polygon_area(p: &[Pt]) -> f64 {
    let n = p.len();
    let mut a = 0.0;
    for i in 0..n { a += p[i].cross(p[(i + 1) % n]); }
    a.abs() / 2.0
}
// Ray-casting point-in-polygon test (returns true for points strictly inside).
fn in_polygon(p: &[Pt], q: Pt) -> bool {
    let n = p.len();
    let mut cnt = 0;
    for i in 0..n {
        let a = p[i];
        let b = p[(i + 1) % n];
        if (a.y <= q.y && q.y < b.y) || (b.y <= q.y && q.y < a.y) {
            let xi = a.x + (b.x - a.x) * (q.y - a.y) / (b.y - a.y);
            if q.x < xi { cnt += 1; }
        }
    }
    cnt & 1 == 1
}
